src/python/pants/backend/python/util_rules/local_dists_pep660.py

```diff
@@ -0,0 +1,539 @@
+        resolve = None
+        # assumption: all owned deps are in the same resolve
+        for dep in owned_deps:
+            if dep.target.has_field(PythonResolveField):
+                resolve = dep.target[PythonResolveField].normalized_value(python_setup)
+                break
+        dists[resolve].append(dist)
+    return ResolveSortedPythonDistributionTargets(
+        FrozenDict({resolve: tuple(targets) for resolve, targets in dists.items()})
+    )
+
+
+@dataclass(frozen=True)
+class EditableLocalDistsRequest:
+    """Request to generate PEP660 wheels of local dists in the given resolve.
+
+    The editable wheel files must not be exported or made available to the end-user (according to
+    PEP 660). Instead, the PEP660 editable wheels serve as intermediate, internal-only,
+    representation of what should be installed in the exported virtualenv to create the editable
+    installs of local python_distributions.
+    """
+
+    resolve: str | None  # None if resolves is not enabled
+
+
+@dataclass(frozen=True)
+class EditableLocalDists:
+    """A Digest populated by editable (PEP660) wheels of local dists.
+
+    According to PEP660, these wheels should not be exported to users and must be discarded
+    after install. Anything that uses this should ensure that these wheels get installed and
+    then deleted.
+
+    Installing PEP660 wheels creates an "editable" install such that the sys.path gets
+    adjusted to include source directories from the build root (not from the sandbox).
+    This is decidedly not hermetic or portable and should only be used locally.
+
+    PEP660 wheels have .dist-info metadata and the .pth files (or similar) that adjust sys.path.
+    """
+
+    optional_digest: Digest | None
+
+
+@rule(desc="Building editable local distributions (PEP 660)")
+async def build_editable_local_dists(  # based on build_local_dists
+    request: EditableLocalDistsRequest,
+    all_dists: ResolveSortedPythonDistributionTargets,
+    python_setup: PythonSetup,
+) -> EditableLocalDists:
+    resolve = request.resolve if python_setup.enable_resolves else None
+    resolve_dists = all_dists.targets.get(resolve, ())
+
+    if not resolve_dists:
+        return EditableLocalDists(None)
+
+    local_dists_wheels = await MultiGet(
+        Get(
+            LocalDistPEP660Wheels,
+            PythonDistributionFieldSet,
+            PythonDistributionFieldSet.create(target),
+        )
+        for target in resolve_dists
+    )
+
+    wheels: list[str] = []
+    wheels_digests = []
+    for local_dist_wheels in local_dists_wheels:
+        wheels.extend(local_dist_wheels.pep660_wheel_paths)
+        wheels_digests.append(local_dist_wheels.pep660_wheels_digest)
+
+    wheels_digest = await Get(Digest, MergeDigests(wheels_digests))
+
+    return EditableLocalDists(wheels_digest)
+
+
+def rules():
+    return (
+        *collect_rules(),
+        *dists_rules(),
+        *setup_py.rules(),
+        *system_binaries.rules(),
```

Comment:
I'm slightly confounded by the extra rule exports here... Until we have a notion of specifying our rule dependencies I think we've just left them out relying on them being registered else where (exception being when they come from other backends that may not be enabled, but in this case it is from the same backend or from core)

Reply:
Yeah. These register functions are a bit of code smell.

It took awhile to get the dependencies right--at least for testing. I also had to reorder the rules in this file or I got rule graph errors about unknown objects. Several of the other util_rules files in the python backend also register other util_rules, so that's what I did here. I would like to see a cleaner registration interface, but I'd rather not mess with it more... rule graph errors are not fun :frowning_face:.

Reply:
Fair enough.